artemis/modules/subdomain_enumeration.py

```diff
@@ -0,0 +1,145 @@
+        for retry_id in range(retries):
+            subdomains = func(domain)
+            if subdomains is not None:
+                return subdomains
+
+            self.log.info("Retrying for domain %s", domain)
+            if retry_id < retries - 1:
+                time.sleep(sleep_time)
+            else:
+                raise UnableToObtainSubdomainsException(f"Unable to obtain subdomains for {domain} after {retries} retries")
+        
+        return set()
+
+    def get_subdomains_from_subfinder(self, domain: str) -> Optional[Set[str]]:
+        subdomains: Set[str] = set()
+        try:
+            result = check_output_log_on_error(
+                [
+                    "subfinder", 
+                    "-d", 
+                    domain, 
+                    "-silent",
+                    "-all",
+                    "-recursive"
+                    ],
+                    self.log
+            )
+            subdomains.update(result.decode().splitlines())
+            print(f"Subdomains found from subfinder: {subdomains}")
+            return subdomains
+        except Exception:
+            self.log.exception("Unable to obtain information from subfinder for domain %s", domain)
+            return None
+
+    def get_subdomains_from_amass(self, domain: str) -> Optional[Set[str]]:
+        subdomains: Set[str] = set()
+        try:
+            result = check_output_log_on_error(
+                [
+                    "amass", #amass is recursive by default use -norecursive to make it non-recursive
+                    "enum",
+                    "-d",
+                    domain,
+                    "-silent"
+                ],
+                self.log
+            )
+            subdomains.update(result.decode().splitlines())
+            return subdomains
+        except Exception as e:
+            if "panic: runtime error: invalid memory address or nil pointer dereference" in str(e):
+                self.log.error(f"Amass encountered a runtime error for domain {domain}. Skipping subdomain enumeration.")
+            else:
+                self.log.exception("Unable to obtain information from amass for domain %s", domain)
+            return None
+
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload("domain")
+        self.log.debug(f"Running subdomain enumeration for domain: {domain}")
+        print(f"Running subdomain enumeration for domain: {domain}")
+
+        try:
+            subdomains = (
+                self.get_subdomains_with_retry(self.get_subdomains_from_subfinder, domain) |
+                self.get_subdomains_with_retry(self.get_subdomains_from_amass, domain)
+            )
+            self.log.debug(f"Subdomains found: {subdomains}")
+            print(f"Subdomains found: {subdomains}")
+        except UnableToObtainSubdomainsException as e:
```

Comment:
this way if subfinder breaks you won't run amass - maybe try...except should be done per-datasource?